### PR TITLE
Metal backend: keep non-packed views in their parent's buffer

### LINKED ISSUE #22983: Metal backend: generated kernels index a packed copy of a non-packed view out of bounds (wrong results, crash in MPSGraph)
### 🐛 Describe the bug

With inductor's default layout optimization, models built on `chunk` + `cat` along C (the YOLO C2f block) return wrong values on the Metal backend, and `yolov8n` crashes inside MPSGraph.

`aoti_torch__reinterpret_tensor` replaces a view that is not densely packed with a packed copy (`materialize_packed`). Kernels generated by inductor never look at the tensor they are handed: they index with the strides they were compiled with, and they write through such views as well as read them. In channels-last a chunk along C is non-packed, and a cat is filled by writing through views of it:

```
buf52 = reinterpret_tensor_wrapper(buf55, 4, {1, 32, 80, 80}, {819200, 1, 10240, 128}, 32)  // alias
```

`buf55` has 819200 elements and the copy that stands in for `buf52` has 204800. The kernel writing "into `buf52`" indexes up to 819200 into that copy, which is an out-of-bounds GPU access, and nothing reaches `buf55`. #22957 lists the lost writes under "Not fixed here" (`linear_chunk_cat_last_dim` is skipped for it); the out-of-bounds part is new, and it is what takes `yolov8n` down: 10 out of 10 runs crash with `EXC_BAD_ACCESS` in `-[MPSGraphExecutable runInternalWithDevice:...]` -> `getFuncOp` -> `mlir::SymbolTable::lookupSymbolIn`, called from `aoti_torch_mps_convolution`, on a pointer that reads as two floats. With `torch._inductor.config.layout_optimization = False` there are no such views and the same model runs correctly.

**Repro.** The C2f block reduced to 1x1 convs, so that it only needs matmuls. For `MODULE_REGISTRY` in `backends/apple/metal/tests/test_modules.py`:

```python
class PointwiseC2f(nn.Module):
    class Inner(nn.Module):
        def __init__(self, channels: int):
            super().__init__()
            self.conv1 = nn.Conv2d(channels, channels, kernel_size=1)
            self.conv2 = nn.Conv2d(channels, channels, kernel_size=1)

        def forward(self, x):
            return x + self.conv2(torch.relu(self.conv1(x)))

    def __init__(self):
        super().__init__()
        self.conv_in = nn.Conv2d(16, 16, kernel_size=1)
        self.inner = nn.ModuleList(PointwiseC2f.Inner(8) for _ in range(2))
        self.conv_out = nn.Conv2d(32, 16, kernel_size=1)

    def forward(self, x):
        parts = list(self.conv_in(x).chunk(2, dim=1))
        for block in self.inner:
            parts.append(block(parts[-1]))
        return self.conv_out(torch.cat(parts, dim=1))


MODULE_REGISTRY["pointwise_c2f"] = {
    "model_class": PointwiseC2f,
    "input_shapes": [(2, 16, 8, 8)],
    "description": "C2f block whose cat is filled through non-packed channels-last views",
}
```

`Output mismatch - max_atol=0.586, max_rtol=1.64` in float32, `0.755 / 1.71` in bfloat16.

### Versions

```
ExecuTorch: #22957 @ 89dbe34077 (main @ 9b91b43098 plus that PR)
PyTorch version: 2.14.0
OS: macOS 27.0 (26A428), arm64, Apple M2 Pro
Clang version: 21.0.0 (clang-2100.3.34.2)
CMake version: 4.4.3
Python version: 3.10.11
```

### PR DESCRIPTION
**Pull request by abdelaziz-mahdy**
### Summary

Fixes #22983

Stacked on #22957, which adds the view registration this builds on; only the last commit (`f669c51dc1`) is new. Once #22957 lands the diff reduces to that commit.

`aoti_torch__reinterpret_tensor` replaced a view that is not densely packed with a packed copy. Kernels generated by inductor index with the strides they were compiled with and write through such views as well as read them, so against a packed copy they lose their writes and index up to `parent numel` into a buffer of `view numel`. Inductor's default channels-last layout makes this the common case: a chunk along C is non-packed there, and a cat is filled through views of it.

- A non-packed view of a Metal buffer now stays where it is and is registered as a view of its parent, like a packed one. The tensor cannot describe such strides (`make_tensor_ptr` rejects them), so it carries packed ones and the real ones are recorded with `metal_record_strided_view`.
- `ETMetalKernelFunction::setArg` takes how the kernel uses the argument (`ArgAccess`). `aoti_torch_mps_set_arg_tensor` passes `kStridedInPlace`: a generated kernel gets the view where it is, which is all it needs.
- Hand-written ops need dense input. `get_mtl_buffer` (mm, addmm, bmm, convolution, topk) and `setArg` with `kRead` hand them a packed copy, made by a small gather kernel encoded on the stream when the op runs. Nothing is committed or waited for, and the op's own encoder stays open. An argument a hand-written kernel writes (`kWrite`) is refused if it is a strided view, and so is a view that cannot be packed.
- An MPSGraph op asked to write its result through such a view, and `aoti_torch_copy_` into one, now fail instead of writing to the wrong place. `aoti_torch_copy_` from one copies through the same gather; the backend uses it to copy a model's outputs out, so a model can return such a view.
- Non-packed views of CPU memory are materialized as before. `materialize_packed` loses its device path, which nothing reaches any more.
- Every handle into memory the runtime owns now holds a count on the allocation it lives in, including views of views and handles copied from views, and gives it back when deleted. Before, a view at an offset never gave its count back (the parent was not freed until cleanup), and nested or copied views took none (the parent could be freed under them). This predates the PR, but the PR sends more views down that path.

This resolves the "Not fixed here" item of #22957.

### Test plan

`pointwise_c2f` added (the YOLO C2f block reduced to 1x1 convs), `linear_chunk_last_dim_output` added (the model's output is such a view), and `linear_chunk_cat_last_dim` enabled; it was skipped for this reason.

```
backends/apple/metal/tests/run_metal_test.sh --build
python -m unittest backends.apple.metal.tests.test_modules.TestMetalBackendModules
```

- Without the runtime change: `pointwise_c2f` fails with `max_atol=0.586` (float32) and `0.755` (bfloat16), `linear_chunk_cat_last_dim` with `max_atol=0.518`.
- With it: `Ran 142 tests`, all passing except the 10 int4 tests that fail on macOS 27.0 for an unrelated reason (#22981, fixed by #22982). `test_metal_memory` 11 of 11 and new `test_metal_strided_view` 6 of 6; each test added for a review finding fails without its change.
- End to end, on a build that also has #22952 and the int4 fix (`Ran 166 tests ... OK`), all-ones input against eager through `executor_runner`: `yolov8n` exported with the default layout went from crashing 10 runs out of 10 to `max abs diff 0.0018` over 5 runs; with `layout_optimization=False` it stays at 0.0019. `yolo11n` with the default layout still stops in `aoti_torch_mps_bmm_out`, which rejects a permuted input (`[2,64,400]`, strides `[1,800,2]`); that is a separate limitation.
- `lintrunner` clean.

### CODE REVIEW

File: backends/apple/metal/runtime/shims/et_metal.mm [hunk 1]
```diff
@@ -505,20 +636,31 @@ int metal_copy_memory(void* dst, const void* src, size_t nbytes, bool src_is_dev
...
         ET_LOG(Error, "ETMetalKernelFunction::setArg: No active encoder");
         return;
     }
 
+    if (!strided_view_in_place) {
+        id<MTLBuffer> packed = metal_packed_copy_of_strided_view(tensor);
+        if (packed) {
+            [encoder_ setBuffer:packed offset:0 atIndex:idx];
+            return;
+        }
```
[the remark this answers is not in the document]
Comment (abdelaziz-mahdy) [reply backends/apple/metal/runtime/shims/et_metal.mm]: Right, fixed in 4648c9a53a. `setArg` runs after `startEncoding()`, and `synchronize()` ends the stream's encoder, so the op was binding into an ended encoder. The packed copy is now a small gather kernel encoded on the stream's current encoder instead of a CPU copy after a `COMMIT_AND_WAIT`: nothing is committed or waited for, it is ordered after the work that fills the view, and `setArg` sets the op's pipeline state again afterwards. The gather binds buffers 28 to 30, so arguments the op already set are left alone. The MPSGraph path (`get_mtl_buffer`) uses the same gather; `executeMPSGraph` ends the encoder and encodes the graph into the same command buffer, so the graph sees the copy. New `test_metal_strided_view` checks the gathered elements; module suite 138 run, only the 10 int4 tests fail (#22981).

File: backends/apple/metal/runtime/shims/memory.cpp [hunk 2]
```diff
@@ -232,12 +232,15 @@ AOTITorchError aoti_torch_delete_tensor_object(AOTITensorHandle tensor) {
...
+  metal_forget_strided_view(tensor);
 
   auto memory_it = memory_to_n_tensor.find(data_ptr);
   if (memory_it != memory_to_n_tensor.end()) {
     int32_t ref_count = memory_it->second;
 
     if (ref_count == NOT_OWN) {
+      // No-op unless this tensor is a view into a Metal buffer.
+      metal_unregister_view(data_ptr);
       tensors.erase(it);
```
[the remark this answers is not in the document]
Comment (abdelaziz-mahdy) [reply backends/apple/metal/runtime/shims/memory.cpp]: Confirmed, fixed in 4648c9a53a. This predates the PR (#22957 has the same path), but this PR sends more views through it. The shims now record which parent's count each offset view took (`view_owner`) and give it back when the view is deleted; the buffer is freed by whichever of view or parent goes last. Two tests in `test_metal_memory` cover both deletion orders: they fail without the change and pass with it (9 of 9).

File: backends/apple/metal/runtime/shims/et_metal.mm [hunk 3]
```diff
@@ -505,20 +609,142 @@ int metal_copy_memory(void* dst, const void* src, size_t nbytes, bool src_is_dev
...
     if (!encoder_) {
         ET_LOG(Error, "ETMetalKernelFunction::setArg: No active encoder");
         return;
     }
 
+    if (!strided_view_in_place) {
+        id<MTLBuffer> packed = encodePackedCopyOfStridedView(encoder_, tensor);
+        if (packed) {
+            [encoder_ setComputePipelineState:cps_];
+            [encoder_ setBuffer:packed offset:0 atIndex:idx];
```
[the remark this answers is not in the document]
Comment (abdelaziz-mahdy) [reply backends/apple/metal/runtime/shims/et_metal.mm]: Fixed in b48b58137b. `setArg` now throws when a strided view cannot be packed, instead of binding it with the packed strides it carries. `FailingToPackIsAnError` in `test_metal_strided_view` fails without the change.

File: backends/apple/metal/runtime/shims/et_metal.mm [hunk 4]
```diff
@@ -505,20 +609,142 @@ int metal_copy_memory(void* dst, const void* src, size_t nbytes, bool src_is_dev
...
+            return;
```
[the remark this answers is not in the document]
Comment (abdelaziz-mahdy) [reply backends/apple/metal/runtime/shims/et_metal.mm]: Fixed in b48b58137b. `setArg` now takes an `ArgAccess`: `kRead` (the default) binds a packed copy, `kWrite` refuses a strided view and throws, since writes to a copy would be lost, and `kStridedInPlace` (generated kernels, via `aoti_torch_mps_set_arg_tensor`) binds the view where it is. The hand-written ops allocate their outputs themselves, so those are never views; they are marked `kWrite` anyway, and so is `gated_delta_rule`'s `state`, the one caller tensor they update in place. Tests: `HandWrittenKernelReadsPackedCopy` and `HandWrittenKernelMayNotWriteAStridedView`.

File: backends/apple/metal/runtime/shims/memory.cpp [hunk 5]
```diff
@@ -232,32 +259,31 @@ AOTITorchError aoti_torch_delete_tensor_object(AOTITensorHandle tensor) {
...
+
+  if (memory_it->second == NOT_OWN) {
+    // No-op unless this tensor is a view into a Metal buffer.
+    metal_unregister_view(data_ptr);
+    // Give back the count the view held on its parent's memory.
+    auto owner = view_owner.find(tensor);
+    if (owner != view_owner.end()) {
+      void* parent = owner->second;
+      view_owner.erase(owner);
+      ET_CHECK_OK_OR_RETURN_ERROR(release_memory(parent));
```
[the remark this answers is not in the document]
Comment (abdelaziz-mahdy) [reply backends/apple/metal/runtime/shims/memory.cpp]: Fixed in b48b58137b. Every handle into owned memory now holds a count on the allocation it lives in: `aoti_torch_new_tensor_handle` takes one through the original's `view_owner` entry, so the allocation is freed only when the last handle is gone. `CopiedViewHandleKeepsParentAlive` deletes the base and the original view first and fails without the change.

File: backends/apple/metal/runtime/shims/memory.cpp [hunk 6]
```diff
@@ -578,19 +625,35 @@ AOTITorchError aoti_torch__reinterpret_tensor(
...
+      // either handle must leave the view registered for the other one.
+      metal_retain_view(data_ptr);
     }
 
     // Increment the reference count for this memory address only if it is owned
     if (memory_to_n_tensor[data_ptr] != NOT_OWN) {
       memory_to_n_tensor[data_ptr] += 1;
+      if (adjusted_data != data_ptr) {
+        // Deleting the view finds `adjusted_data`, not `data_ptr`; remember
+        // where its count went.
```
[the remark this answers is not in the document]
Comment (abdelaziz-mahdy) [reply backends/apple/metal/runtime/shims/memory.cpp]: Fixed in b48b58137b. `aoti_torch__reinterpret_tensor` now finds the allocation through `self`'s `view_owner` entry when `self` is itself a view, and the new view holds a count on it. `NestedViewKeepsParentAlive` covers nested offsets 0 and 2, deleting the base and the first view before the nested one; it fails without the change.

File: backends/apple/metal/tests/test_modules.py [hunk 7]
```diff
@@ -278,6 +278,157 @@ def forward(self, x: torch.Tensor):
...
 
 
+# -------------------------------------------------------------------------
+# Views with a storage offset. The chunks below are views into the first
+# linear's output; the second chunk starts partway into that buffer. The cat
+# variants also write through such views.
+# -------------------------------------------------------------------------
+class LinearChunkLastDim(nn.Module):
+    """Chunking the last dim gives a non-packed view (its row stride is still
+    the parent's), which reinterpret_tensor has to materialize."""
```
[the remark this answers is not in the document]
Comment (abdelaziz-mahdy) [reply backends/apple/metal/tests/test_modules.py]: Updated in b48b58137b, along with the `NarrowLastDim` comment, which said the same thing.

**Comment by abdelaziz-mahdy** [comment]
Pushed b48b58137b for the five review comments above. Going over the rest of the change I also found a regression the reviews did not cover: 4648c9a53a made `aoti_torch_copy_` refuse non-packed views, and the backend copies a model's outputs out with it, so a model returning such a view (e.g. `linear(x).chunk(2, dim=-1)[1]`) failed with `aoti_torch_copy_ does not support views that are not densely packed`. A strided source is now copied through the same GPU gather, and `linear_chunk_last_dim_output` covers it; it fails on 4648c9a53a and passes now.

**Comment by abdelaziz-mahdy** [comment]
Merged the updated #22957 (67d8d5dcb5) into this branch as 878aa22fa8. The ownership accounting that was here now comes from #22957. Two things needed changing for the two to fit together:

- #22957 gives a packed offset view of CPU memory its own no-copy buffer again. Such a view counts as a device pointer, but it cannot hold a non-packed view in place, since its buffer covers only itself. A non-packed view of it is now materialized, like any view of CPU memory (`NonPackedViewOfCpuBackedViewIsMaterialized`).
- `materialize_packed` waits for the GPU again before reading: kernels can write CPU memory through those buffers, and a copy made afterwards has to see the writes (`MaterializingCpuMemoryWaitsForGpuWrites`).

Both tests fail without the change. `test_metal_memory` 16/16, `test_metal_strided_view` 8/8, module suite 142 run with only the 10 int4 failures (#22982, merged, not in this branch), `lintrunner` clean.

File: backends/apple/metal/runtime/test/test_strided_view.mm [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include <gtest/gtest.h>
+
+#include <cstdint>
+#include <cstring>
+#include <stdexcept>
+#include <vector>
```
[the remark this answers is not in the document]
Comment (abdelaziz-mahdy) [reply backends/apple/metal/runtime/test/test_strided_view.mm:16]: Added in ca9447bd0e. I also went through the other files this PR touches for the same thing: `test_strided_view.mm` now includes `<memory>` as well, `et_metal.mm` `<cstdint>`, `<stdexcept>` and `<string>` for the gather and `setArg` code, and `common.mm` `<cstring>` for `std::strcmp`. The same push brings in #22957's host/device copy ordering fix (090bf1aa9c). `test_metal_memory` 18/18, `test_metal_strided_view` 8/8, module suite 142 run with only the 10 int4 failures (#22982 not in this branch), `lintrunner` clean.

**Comment by abdelaziz-mahdy** [comment]
Merged the updated #22957 as 2378f594b1, which also brings in main (#22952 and #22982). The one conflict was in `materialize_packed`, which both branches now make wait unconditionally before copying. I kept #22957's comment, which also covers the recycled destination buffer.

On how this fits with #22952: the tensor for a non-packed view carries packed strides, so `is_row_major_dense` accepts it in `bmm` and the 4-bit linear, and they get the packed copy (a written argument is still refused). I also checked whether a convolution could be handed such a view and then pick the wrong output layout. It cannot come from generated code: for `conv(chunk(conv(x), 2, dim=1)[1])` inductor copies the chunk into a dense buffer before the extern convolution.

`test_metal_memory` 22/22, `test_metal_strided_view` 8/8, module suite `Ran 178 tests ... OK`, no failures or skips (the int4 tests pass with #22982). `lintrunner` clean.
